File: src/semp_workflow/modules/rdp_qb.py
"""Queue Binding module - rdp_qb.add, rdp_qb.delete."""

from __future__ import annotations

import logging

from ..exceptions import SEMPError
from ..models import ActionResult, ResultStatus
from ..semp.client import SempClient, ALREADY_EXISTS
from ..semp.helpers import check_name_length, clean_payload, coerce_bool, enc
from .base import BaseModule

logger = logging.getLogger(__name__)


class QueueBindingAdd(BaseModule):
    description = "Bind a queue to an RDP for message delivery. Skipped if the binding already exists."
    params = {
        "restDeliveryPointName":              {"type": "string",  "required": True,  "description": "Name of the REST Delivery Point"},
        "queueBindingName":                   {"type": "string",  "required": True,  "description": "Name of the queue to bind"},
        "postRequestTarget":                  {"type": "string",  "required": False, "description": "HTTP request target path appended to the REST consumer URL"},
        "gatewayReplaceTargetAuthorityEnabled":{"type": "boolean", "required": False, "description": "Replace the authority in forwarded HTTP requests with the remote host"},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        rdp_name = args.get("restDeliveryPointName", "")
        binding_name = args.get("queueBindingName", "")
        if not rdp_name or not binding_name:
            return ActionResult(
                ResultStatus.FAILED, "Missing required args: restDeliveryPointName, queueBindingName"
            )

        for field, value in (("restDeliveryPointName", rdp_name), ("queueBindingName", binding_name)):
            if err := check_name_length(field, value):
                return ActionResult(ResultStatus.FAILED, err)

        path = f"restDeliveryPoints/{enc(rdp_name)}/queueBindings/{enc(binding_name)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking queue binding: {e}")

        if exists:
            return ActionResult(
                ResultStatus.SKIPPED,
                f"Queue binding '{binding_name}' already exists on RDP '{rdp_name}'",
            )

        if dry_run:
            return ActionResult(
                ResultStatus.DRYRUN,
                f"Would create queue binding '{binding_name}' on RDP '{rdp_name}'",
            )

        try:
            payload = clean_payload(args)
            payload.pop("restDeliveryPointName", None)  # path param, not a body field
            if "gatewayReplaceTargetAuthorityEnabled" in payload:
                payload["gatewayReplaceTargetAuthorityEnabled"] = coerce_bool(
                    payload["gatewayReplaceTargetAuthorityEnabled"]
                )
            client.create(
                f"restDeliveryPoints/{enc(rdp_name)}/queueBindings", payload
            )
            return ActionResult(
                ResultStatus.OK,
                f"Queue binding '{binding_name}' created on RDP '{rdp_name}'",            )
        except SEMPError as e:
            if e.semp_code == ALREADY_EXISTS:
                return ActionResult(
                    ResultStatus.SKIPPED,
                    f"Queue binding '{binding_name}' already exists on RDP '{rdp_name}'",
                )
            return ActionResult(
                ResultStatus.FAILED, f"Failed to create queue binding: {e}"
            )


class QueueBindingDelete(BaseModule):
    description = "Remove a queue binding from an RDP. Skipped if the binding does not exist."
    params = {
        "restDeliveryPointName": {"type": "string", "required": True, "description": "Name of the REST Delivery Point"},
        "queueBindingName":      {"type": "string", "required": True, "description": "Name of the bound queue to remove"},
    }

    def execute(self, client: SempClient, args: dict, dry_run: bool = False) -> ActionResult:
        rdp_name = args.get("restDeliveryPointName", "")
        binding_name = args.get("queueBindingName", "")
        if not rdp_name or not binding_name:
            return ActionResult(
                ResultStatus.FAILED, "Missing required args: restDeliveryPointName, queueBindingName"
            )

        for field, value in (("restDeliveryPointName", rdp_name), ("queueBindingName", binding_name)):
            if err := check_name_length(field, value):
                return ActionResult(ResultStatus.FAILED, err)

        path = f"restDeliveryPoints/{enc(rdp_name)}/queueBindings/{enc(binding_name)}"

        try:
            exists, _ = client.exists(path)
        except SEMPError as e:
            return ActionResult(ResultStatus.FAILED, f"Error checking queue binding: {e}")

        if not exists:
            return ActionResult(
                ResultStatus.SKIPPED,
                f"Queue binding '{binding_name}' does not exist on RDP '{rdp_name}'",
            )

        if dry_run:
            return ActionResult(
                ResultStatus.DRYRUN,
                f"Would delete queue binding '{binding_name}' from RDP '{rdp_name}'",
            )

        try:
            client.delete(path)
            return ActionResult(
                ResultStatus.OK,
                f"Queue binding '{binding_name}' deleted from RDP '{rdp_name}'",            )
        except SEMPError as e:
            return ActionResult(
                ResultStatus.FAILED, f"Failed to delete queue binding: {e}"
            )


MODULES = {
    "rdp_qb.add": QueueBindingAdd,
    "rdp_qb.delete": QueueBindingDelete,
}

File: internal/modules/rdp_qb.go
package modules

import (
	"fmt"

	"semp-workflow/internal/models"
	"semp-workflow/internal/semp"
)

func init() {
	register("rdp_qb.add", queueBindingAdd{})
	register("rdp_qb.delete", queueBindingDelete{})
}

type queueBindingAdd struct{}

func (queueBindingAdd) Description() string {
	return "Bind a queue to an RDP for message delivery. Skipped if the binding already exists."
}

func (queueBindingAdd) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "restDeliveryPointName", Type: "string", Required: true, Description: "Name of the REST Delivery Point"},
		{Name: "queueBindingName", Type: "string", Required: true, Description: "Name of the queue to bind"},
		{Name: "postRequestTarget", Type: "string", Description: "HTTP request target path appended to the REST consumer URL"},
		{Name: "gatewayReplaceTargetAuthorityEnabled", Type: "boolean", Description: "Replace the authority in forwarded HTTP requests with the remote host"},
	}
}

func (queueBindingAdd) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	rdpName := argStr(args, "restDeliveryPointName", "")
	bindingName := argStr(args, "queueBindingName", "")
	if rdpName == "" || bindingName == "" {
		return failed("Missing required args: restDeliveryPointName, queueBindingName")
	}
	for _, f := range []struct{ field, value string }{
		{"restDeliveryPointName", rdpName},
		{"queueBindingName", bindingName},
	} {
		if err := semp.CheckNameLength(f.field, f.value); err != "" {
			return failed(err)
		}
	}

	path := fmt.Sprintf("restDeliveryPoints/%s/queueBindings/%s", semp.Enc(rdpName), semp.Enc(bindingName))

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking queue binding: %s", err))
	}
	if exists {
		return skipped(fmt.Sprintf("Queue binding '%s' already exists on RDP '%s'", bindingName, rdpName))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would create queue binding '%s' on RDP '%s'", bindingName, rdpName))
	}

	payload := semp.CleanPayload(args)
	delete(payload, "restDeliveryPointName") // path param, not a body field
	coerceBoolFields(payload, "gatewayReplaceTargetAuthorityEnabled")
	if _, err := c.Create(fmt.Sprintf("restDeliveryPoints/%s/queueBindings", semp.Enc(rdpName)), payload); err != nil {
		if sempCodeOf(err) == semp.AlreadyExists {
			return skipped(fmt.Sprintf("Queue binding '%s' already exists on RDP '%s'", bindingName, rdpName))
		}
		return failed(fmt.Sprintf("Failed to create queue binding: %s", err))
	}
	return ok(fmt.Sprintf("Queue binding '%s' created on RDP '%s'", bindingName, rdpName))
}

type queueBindingDelete struct{}

func (queueBindingDelete) Description() string {
	return "Remove a queue binding from an RDP. Skipped if the binding does not exist."
}

func (queueBindingDelete) Params() []ParamSpec {
	return []ParamSpec{
		{Name: "restDeliveryPointName", Type: "string", Required: true, Description: "Name of the REST Delivery Point"},
		{Name: "queueBindingName", Type: "string", Required: true, Description: "Name of the bound queue to remove"},
	}
}

func (queueBindingDelete) Execute(c Client, args map[string]any, dryRun bool) models.ActionResult {
	rdpName := argStr(args, "restDeliveryPointName", "")
	bindingName := argStr(args, "queueBindingName", "")
	if rdpName == "" || bindingName == "" {
		return failed("Missing required args: restDeliveryPointName, queueBindingName")
	}
	for _, f := range []struct{ field, value string }{
		{"restDeliveryPointName", rdpName},
		{"queueBindingName", bindingName},
	} {
		if err := semp.CheckNameLength(f.field, f.value); err != "" {
			return failed(err)
		}
	}

	path := fmt.Sprintf("restDeliveryPoints/%s/queueBindings/%s", semp.Enc(rdpName), semp.Enc(bindingName))

	exists, _, err := c.Exists(path)
	if err != nil {
		return failed(fmt.Sprintf("Error checking queue binding: %s", err))
	}
	if !exists {
		return skipped(fmt.Sprintf("Queue binding '%s' does not exist on RDP '%s'", bindingName, rdpName))
	}
	if dryRun {
		return dryrun(fmt.Sprintf("Would delete queue binding '%s' from RDP '%s'", bindingName, rdpName))
	}

	if err := c.Delete(path); err != nil {
		return failed(fmt.Sprintf("Failed to delete queue binding: %s", err))
	}
	return ok(fmt.Sprintf("Queue binding '%s' deleted from RDP '%s'", bindingName, rdpName))
}
